node/src/store/header_ranges.rs

```diff
@@ -0,0 +1,560 @@
+        {
+            let mut prev: Option<&RangeInclusive<u64>> = None;
+
+            for range in &from {
+                assert!(
+                    range.start() <= range.end(),
+                    "range isn't sorted internally"
+                );
+
+                if let Some(prev) = prev {
+                    assert!(
+                        prev.end() < range.start(),
+                        "header ranges aren't sorted correctly"
+                    );
+                }
+
+                prev = Some(range);
+            }
+        }
+
+        Self(from)
+    }
+}
+
+impl Display for HeaderRanges {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "[")?;
+        for (idx, range) in self.0.iter().enumerate() {
+            if idx == 0 {
+                write!(f, "{}-{}", range.start(), range.end())?;
+            } else {
+                write!(f, ", {}-{}", range.start(), range.end())?;
+            }
+        }
+        write!(f, "]")
+    }
+}
+
+impl AsRef<[RangeInclusive<u64>]> for HeaderRanges {
+    fn as_ref(&self) -> &[RangeInclusive<u64>] {
+        &self.0
+    }
+}
+
+impl IntoIterator for HeaderRanges {
+    type Item = u64;
+    type IntoIter = HeaderRangesIterator;
+    fn into_iter(self) -> Self::IntoIter {
+        let mut outer_iter = self.0.into_iter();
+        HeaderRangesIterator {
+            inner_iter: outer_iter.next(),
+            outer_iter,
+        }
+    }
+}
+
+pub struct HeaderRangesIterator {
+    inner_iter: Option<HeaderRange>,
+    outer_iter: IntoIter<[HeaderRange; 2]>,
+}
+
+impl HeaderRangesIterator {
+    pub fn next_batch(&mut self, limit: u64) -> Option<HeaderRange> {
+        let current_range = self.inner_iter.take()?;
+
+        if current_range.len() <= limit {
+            self.inner_iter = self.outer_iter.next();
+            Some(current_range)
+        } else {
+            let returned_range = *current_range.start()..=*current_range.start() + limit - 1;
+            self.inner_iter = Some(*current_range.start() + limit..=*current_range.end());
+            Some(returned_range)
+        }
+    }
+}
+
+impl Iterator for HeaderRangesIterator {
+    type Item = u64;
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            if let Some(v) = self.inner_iter.as_mut()?.next() {
+                return Some(v);
+            }
+            self.inner_iter = self.outer_iter.next();
+        }
+    }
+}
+
+/// Extends test header generator for easier insertion into the store
+pub trait ExtendedHeaderGeneratorExt {
+    /// Generate next amount verified headers
+    fn next_many_verified(&mut self, amount: u64) -> VerifiedExtendedHeaders;
+}
+
+#[cfg(any(test, feature = "test-utils"))]
+impl ExtendedHeaderGeneratorExt for ExtendedHeaderGenerator {
+    fn next_many_verified(&mut self, amount: u64) -> VerifiedExtendedHeaders {
+        VerifiedExtendedHeaders::from_verified_vec(self.next_many(amount))
+    }
+}
```

Comment:
Maybe for later: We can actually move `VerifiedExtendedHeaders` in `celestia-types` and `ExtendedHeaderGenerator` should always return `VerifiedExtendedHeaders`.

Reply:
can there be a case we want to generate not verified?

Reply:
No. But we have this [helper](https://github.com/eigerco/lumina/blob/a20665fc42e73ef35f4b497130f3006e755e1ca1/types/src/test_utils.rs#L275)

Reply:
https://github.com/eigerco/lumina/issues/301